### PR TITLE
Add OKX adapter

### PR DESCRIPTION
**Pull request by miller-moore**
# Pull Request

Include a summary of the changes.

## Adds an OKX adapter

Delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How has this change been tested?

The adapter has only been tested by contrived live experiments with linear perpetual instruments. No code tests have yet been implemented.

### CODE REVIEW

**Comment by cjdsellers** [comment]
Hey @miller-moore

Many thanks for the contribution! On initial glance its looking good, but I'll have a closer look tomorrow.

**Comment by cjdsellers** [comment]
@miller-moore

Solid work here, what's the current plan then? I understand you implemented mostly for the perp derivatives - was there anything else you intended to add/improve before this is merged? we could then just iterate and improve from there.

File: nautilus_trader/adapters/okx/common/enums.py [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+    def parse_nautilus_order_side(self, order_side: OrderSide) -> OKXOrderSide:
+        return check_dict_keys(order_side, self.nautilus_to_okx_order_side)
+
+    def parse_nautilus_trigger_type(self, trigger_type: TriggerType) -> OKXTriggerType:
+        return check_dict_keys(trigger_type, self.nautilus_to_okx_trigger_type)
+
+    def parse_okx_trigger_type(self, trigger_type: OKXTriggerType) -> TriggerType:
+        return check_dict_keys(trigger_type, self.okx_to_nautilus_trigger_type)
+
+    def parse_okx_order_type(self, ordType: OKXOrderType) -> OrderType:
```
[the remark this answers is not in the document]
Comment (miller-moore) [reply nautilus_trader/adapters/okx/common/enums.py:448]: Indeed and to fully implement OKX’s algo order capabilities (tp, sl, trailing, contingencies, etc) will require much improved parsing logic.

File: nautilus_trader/adapters/okx/common/symbol.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+from nautilus_trader.adapters.okx.common.enums import OKXInstrumentType
+from nautilus_trader.core.correctness import PyCondition
+from nautilus_trader.model.identifiers import InstrumentId
+from nautilus_trader.model.identifiers import Symbol
+
+
+VALID_SUFFIXES: Final[list[str]] = [
+    "-SPOT",
+    "-MARGIN",
+    "-LINEAR",
```
[the remark this answers is not in the document]
Comment (miller-moore) [reply nautilus_trader/adapters/okx/common/symbol.py:14]: No objection to your suggestion. One thing to keep in mind is that OKX linear and inverse perps already contain the suffix “-SWAP” on their internal symbols. I kept the Bybit convention to distinguish them explicitly but a simpler alternative is always desirable

File: nautilus_trader/adapters/okx/config.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        If the client is connecting to the OKX demo API.
+
+    """
+
+    api_key: str | None = None
+    api_secret: str | None = None
+    passphrase: str | None = None
+    instrument_types: tuple[OKXInstrumentType] | None = (OKXInstrumentType.SWAP,)
+    contract_types: tuple[OKXContractType] | None = (OKXContractType.LINEAR,)
+    base_url_http: str | None = get_http_base_url()
```
[the remark this answers is not in the document]
Comment (miller-moore) [reply nautilus_trader/adapters/okx/config.py:64]: Sounds good. I’ll refactor accordingly

**Comment by miller-moore** [comment]
> @miller-moore
> 
> Solid work here, what's the current plan then? I understand you implemented mostly for the perp derivatives - was there anything else you intended to add/improve before this is merged? we could then just iterate and improve from there.

It would be ideal to contribute the spot and margin product types and whatever adapter changes are needed to support them but my bandwidth for such a contribution does not exist at this time. 

There is one odd hack or idiosyncrasy I added to prevent OKX websocket connections from being severed, which is an async task that checks for the most recent last pong time received by the websocket and, if the time delta exceeds the allowed time delta (currently hardcoded to 1 minute), a completely new `WebSocketClient`  is created and connected. Awkwardly, the `OKXWebsocketClient` attribute `_last_pong` (`pandas.Timestamp`) is updated in the respective websocket handlers in `OKXDataClient` and `OKXExecutionClient`. Furthermore, for reasons unknown to me, the task is not properly cancelled by the nautilus loop signal handler when the process is killed. I don’t know if this pong-based reconnect hack was ultimately necessary but the conditions for not receiving pongs on time was met frequently enough in live testing that I felt it should be retained (eg, maybe once per hour but sometimes more and sometimes less). All keeping in mind, each `WebSocketClient` is configured with a 20 second ping heartbeat.

Otherwise, iterate and improve is my best recommendation

**Comment by cjdsellers** [comment]
Understood, that is strange on the ping-pong issue. Was there anything else you wanted to add/change before we merge this then?

**Comment by miller-moore** [comment]
> Understood, that is strange on the ping-pong issue. Was there anything else you wanted to add/change before we merge this then?

Nope!

**Pull request merged**